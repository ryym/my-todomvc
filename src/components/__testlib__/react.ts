import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

// Though we used @testing-library/react first, it looks like it has a bug
// about the change event simulation (https://github.com/testing-library/react-testing-library/issues/359).
// And the official react-dom/test-utils is useful enough.

export const render = (element: ReactElement<any>) => {
  const container = document.createElement('div');
  act(() => {
    ReactDOM.render(element, container);
  });
  return container;
};

export const getByTestId = (container: HTMLElement, testId: string): HTMLElement => {
  const element = container.querySelector(`[data-testid="${testId}"]`) as HTMLElement;
  if (element == null) {
    throw new Error(`Test ID not found: ${testId}`);
  }
  return element;
};

export const renderSnapshot = (element: ReactElement<any>) => {
  return renderer.create(element).toJSON();
};
