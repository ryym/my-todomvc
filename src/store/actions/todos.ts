import { defineActions, effect } from 'redy';

let _globalTodoId = 1;

export const $todos = defineActions('todos', {
  AddNew: effect((title: string) => async dispatch => {
    if (title != null) {
      title = title.trim();
      if (title.length > 0) {
        dispatch($todos.Add(_globalTodoId++, title));
      }
    }
  }),

  Add: (id: number, title: string) => ({ id, title }),

  Remove: (id: number) => ({ id }),

  ChangeTitle: (id: number, title: string) => ({ id, title }),

  ToggleCompleted: (id: number, completed: boolean) => ({ id, completed }),

  ToggleCompletedAll: (completed: boolean) => ({ completed }),

  ClearCompleted: () => {},
});
