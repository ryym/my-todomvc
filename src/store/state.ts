import { Todos } from '../lib/Todos';
import { TodoState } from '../lib/TodoState';

export interface State {
  readonly todos: TodosState;
  readonly filters: FiltersState;
}

export type TodosState = Todos;

export interface FiltersState {
  readonly todoState: TodoState.Value;
}

export const initFilters = (): FiltersState => ({
  todoState: TodoState.Value.All,
});
