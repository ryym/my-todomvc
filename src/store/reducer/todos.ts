import { defineReducer, on } from 'redy';
import { $todos } from '../actions/todos';
import { Todos } from '../../lib/Todos';

export const reduceTodos = defineReducer(Todos.init(), [
  on($todos.Add, (todos, { id, title }) => {
    const newTodo = Todos.create(id, title);
    return Todos.add(todos, newTodo);
  }),

  on($todos.Remove, (todos, { id }) => {
    return Todos.remove(todos, [id]);
  }),

  on($todos.ChangeTitle, (todos, { id, title }) => {
    const todo = Todos.find(todos, id);
    const nextTodo = { ...todo, title };
    return Todos.replace(todos, [nextTodo]);
  }),

  on($todos.ToggleCompleted, (todos, { id, completed }) => {
    const todo = Todos.find(todos, id);
    return Todos.replace(todos, [{ ...todo, completed }]);
  }),

  on($todos.ToggleCompletedAll, (todos, { completed }) => {
    const toggledTodos = todos.map(todo => ({ ...todo, completed }));
    return Todos.replace(todos, toggledTodos);
  }),

  on($todos.ClearCompleted, todos => {
    const completedIds = todos.filter(todo => todo.completed).map(t => t.id);
    return Todos.remove(todos, completedIds);
  }),
]);
