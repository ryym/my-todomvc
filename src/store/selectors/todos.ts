import { State } from '../state';
import { Todo } from '../../lib/Todos';
import { TodoState } from '../../lib/TodoState';
import { unreachable } from '../../lib/unreachable';

export const selectFilteredTodos = ({ todos, filters }: State): Todo[] => {
  switch (filters.todoState) {
    case TodoState.Value.All:
      return todos;

    case TodoState.Value.Active:
      return todos.filter(t => !t.completed);

    case TodoState.Value.Completed:
      return todos.filter(t => t.completed);

    default:
      return unreachable(filters.todoState);
  }
};
