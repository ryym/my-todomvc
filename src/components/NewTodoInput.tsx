import React, { useState } from 'react';
import { connect } from '../store/connect';
import { $todos } from '../store/actions/todos';

export const NewTodoInput = connect(
  () => ({}),

  function NewTodoInput({ dispatch }) {
    const [title, setTitle] = useState('');

    return (
      <input
        data-testid="new-todo"
        type="text"
        className="new-todo"
        placeholder="What needs to be done?"
        autoFocus
        value={title}
        onChange={event => setTitle(event.currentTarget.value)}
        onKeyDown={event => {
          if (event.key === 'Enter') {
            dispatch($todos.AddNew(title));
            setTitle('');
          }
        }}
      />
    );
  }
);
