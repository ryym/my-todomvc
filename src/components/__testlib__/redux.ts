import { Dispatch, Store, createStore } from 'redux';
import { State } from '../../store/state';

export function mockStore<S = State>(initialState: S, dispatch?: Dispatch): Store<S> {
  const noopReducer = (state: any) => state;
  let mockStore = createStore(noopReducer, initialState);
  if (dispatch) {
    mockStore = { ...mockStore, dispatch };
  }
  return mockStore;
}
