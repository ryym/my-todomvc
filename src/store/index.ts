import { createStore, applyMiddleware, compose, Store } from 'redux';
import { redyMiddleware } from 'redy';
import logger from 'redux-logger';
import { createReducer } from './reducer';
import { State } from './state';

// https://github.com/zalmoxisus/redux-devtools-extension#installation
const composeEnhancers: typeof compose =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const configureStore = (initialState: Partial<State>): Store<State> => {
  const reducer = createReducer();
  const enhancer = composeEnhancers(applyMiddleware(logger, redyMiddleware()));
  return createStore(reducer, initialState, enhancer);
};
