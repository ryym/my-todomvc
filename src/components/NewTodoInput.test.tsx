import React from 'react';
import { spy } from 'sinon';
import { Provider } from 'react-redux';
import { act, Simulate } from 'react-dom/test-utils';
import { renderSnapshot, render, getByTestId } from './__testlib__/react';
import { mockStore } from './__testlib__/redux';
import { initState } from '../store/state';
import { $todos } from '../store/actions/todos';
import { NewTodoInput } from './NewTodoInput';

it('matches snapshot', () => {
  const store = mockStore(initState());
  const tree = renderSnapshot(
    <Provider store={store}>
      <NewTodoInput />
    </Provider>
  );
  expect(tree).toMatchSnapshot();
});

it('adds new todo when Enter is pressed', () => {
  const dispatch = spy();
  const container = render(<NewTodoInput.WrappedComponent dispatch={dispatch} />);
  const input = getByTestId(container, 'new-todo') as HTMLInputElement;

  act(() => {
    input.value = 'Buy Switch';
    Simulate.change(input);
  });

  act(() => {
    Simulate.keyDown(input, { key: 'Enter' });
  });

  expect(dispatch.args).toEqual([[$todos.AddNew('Buy Switch')]]);
});
