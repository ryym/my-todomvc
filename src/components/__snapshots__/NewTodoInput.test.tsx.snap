// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<input
  autoFocus={true}
  className="new-todo"
  data-testid="new-todo"
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="What needs to be done?"
  type="text"
  value=""
/>
`;
